### PR TITLE
ref(codeowners): Update on_delete for FK in ProjectCodeOwners table

### PR DESCRIPTION
**Pull request by MeredithAnya**
**Context:**
The `RepositoryProjectPathConfig` table is what we refer to in the product as **Code Mappings**. 

These mappings were [originally introduced](https://github.com/getsentry/sentry/pull/21331) to support Stack Trace Linking, but they are also required for the Code Owners feature ecosystem is working on. These code mappings were added as a FK in https://github.com/getsentry/sentry/pull/23715.

**Deleting a Code Mapping:**
Because the code owners _needs_ a code mapping, we don't want to let users delete a code mapping if it's currently tied to a `ProjectCodeOwner` record. 

_problem_: The `FlexibleForeignKey` in Sentry defaults to `models.CASCADE` so right now if you a delete the code mapping connected to a code owner record, we will just delete both.
https://github.com/getsentry/sentry/blob/d4426e90b66eca8f352697dcb097b81f86a33da4/src/sentry/db/models/fields/foreignkey.py#L7-L10

_solution_: Add a migration to add `models.PROTECT` to prevent this from happening. Once https://github.com/getsentry/sentry/pull/23855 is merged we can add a try/except for the `ProtectedError` on the `DELETE` endpoint.

### CODE REVIEW

File: src/sentry/migrations/0164_add_protect_on_delete_codeowners.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+    operations = [
+        migrations.AlterField(
+            model_name="projectcodeowners",
+            name="repository_project_path_config",
+            field=sentry.db.models.fields.foreignkey.FlexibleForeignKey(
+                on_delete=django.db.models.deletion.PROTECT, to="sentry.RepositoryProjectPathConfig"
+            ),
+        ),
+    ]
```
Comment (wedamija) [inline src/sentry/migrations/0164_add_protect_on_delete_codeowners.py:41]: Django is unnecessarily recreating the fk here, lets change this to

```
    operations = [
        migrations.SeparateDatabaseAndState(state_operations=[
            migrations.AlterField(
                model_name="projectcodeowners",
                name="repository_project_path_config",
                field=sentry.db.models.fields.foreignkey.FlexibleForeignKey(
                    on_delete=django.db.models.deletion.PROTECT, to="sentry.RepositoryProjectPathConfig"
                ),
            ),
        ])
    ]
```

**Review by wedamija — APPROVED** [review]

**Review by wedamija — COMMENTED** [review]
Btw, we missed this earlier, but we should probably add `ProjectCodeOwners` to https://github.com/getsentry/sentry/blob/658504a9b2cab9b2c2578cd2e0f474817819a7f9/src/sentry/deletions/defaults/project.py#L17

**Comment by MeredithAnya** [comment]
> Btw, we missed this earlier, but we should probably add `ProjectCodeOwners` to
> 
> https://github.com/getsentry/sentry/blob/658504a9b2cab9b2c2578cd2e0f474817819a7f9/src/sentry/deletions/defaults/project.py#L17

@wedamija I actually do wanna ask about that cause we probably need that for `RepositoryProjectPathConfig` as well but with this constraint added, is there way to make sure that the `ProjectCodeOwner` gets deleted first? This is going to be similar to if an integration is uninstalled as well (the `organization_integration` is deleted).

**Comment by wedamija** [comment]
> > Btw, we missed this earlier, but we should probably add `ProjectCodeOwners` to
> > https://github.com/getsentry/sentry/blob/658504a9b2cab9b2c2578cd2e0f474817819a7f9/src/sentry/deletions/defaults/project.py#L17
> 
> @wedamija I actually do wanna ask about that cause we probably need that for `RepositoryProjectPathConfig` as well but with this constraint added, is there way to make sure that the `ProjectCodeOwner` gets deleted first? This is going to be similar to if an integration is uninstalled as well (the `organization_integration` is deleted) cause we also will need the `ProjectCodeOwner` to be deleted prior to the `RepositoryProjectPathConfig` being deleted

Reading the code, it looks like these relations are processed in order. So I think you can just make sure they're ordered correctly in the list, and add a comment explaining that the order is important.

**Pull request merged**